cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh

```diff
@@ -45,67 +49,105 @@ constexpr static inline const float kAdjustCentersWeight = 7.0f;
-
-  float alpha;
-  float beta;
-  switch (metric) {
-    case raft::distance::DistanceType::InnerProduct: {
-      alpha = -1.0;
-      beta  = 0.0;
-    } break;
-    case raft::distance::DistanceType::L2Expanded:
-    case raft::distance::DistanceType::L2Unexpanded: {
-      rmm::device_uvector<float> sqsum_centers(n_clusters, stream, mr);
-      rmm::device_uvector<float> sqsum_data(n_rows, stream, mr);
-      utils::dots_along_rows(n_clusters, dim, centers, sqsum_centers.data(), stream);
-      utils::dots_along_rows(n_rows, dim, dataset, sqsum_data.data(), stream);
-      utils::outer_add(
-        sqsum_data.data(), n_rows, sqsum_centers.data(), n_clusters, distances.data(), stream);
-      alpha = -2.0;
-      beta  = 1.0;
-    } break;
-    // NB: update the description of `knn::ivf_flat::build` when adding here a new metric.
-    default: RAFT_FAIL("The chosen distance metric is not supported (%d)", int(metric));
+  if (metric == raft::distance::DistanceType::L2Expanded ||
+      metric == raft::distance::DistanceType::L2SqrtExpanded) {
+    // todo(lsugy): pass
+    auto workspace = raft::make_device_vector<char, IdxT>(handle, (sizeof(int)) * n_rows);
+
+    raft::distance::MinAndDistanceReduceOp<LabelT, float> redOp;
+    raft::distance::KVPMinReduce<LabelT, float> pairRedOp;
+
+    auto minClusterAndDistance =
+      raft::make_device_vector<cub::KeyValuePair<LabelT, float>, IdxT>(handle, n_rows);
+    cub::KeyValuePair<LabelT, float> initial_value(0, std::numeric_limits<float>::max());
+    thrust::fill(handle.get_thrust_policy(),
+                 minClusterAndDistance.data_handle(),
+                 minClusterAndDistance.data_handle() + minClusterAndDistance.size(),
+                 initial_value);
+
+    auto centroidsNorm = raft::make_device_vector<float, IdxT>(handle, n_clusters);
+    raft::linalg::rowNorm<float, IdxT>(
+      centroidsNorm.data_handle(), centers, dim, n_clusters, raft::linalg::L2Norm, true, stream);
+
+    // todo(lsugy): should fusedL2NN take IdxT!=LabelT?
+    raft::distance::fusedL2NN<float, cub::KeyValuePair<LabelT, float>, LabelT>(
+      minClusterAndDistance.data_handle(),
+      dataset,
+      centers,
+      dataset_norm,
+      centroidsNorm.data_handle(),
+      n_rows,
+      n_clusters,
+      dim,
+      (void*)workspace.data_handle(),
+      redOp,
+      pairRedOp,
+      (metric == raft::distance::DistanceType::L2Expanded) ? false : true,
+      false,
+      stream);
+
+    // todo(lsugy): this is temporary!
+    // Copy keys to output labels
+    thrust::transform(handle.get_thrust_policy(),
+                      minClusterAndDistance.data_handle(),
+                      minClusterAndDistance.data_handle() + n_rows,
+                      labels,
+                      [=] __device__(cub::KeyValuePair<LabelT, float> kvp) { return kvp.key; });
+
+    // todo(lsugy): return distances!
+  } else if (metric == raft::distance::DistanceType::InnerProduct) {
```

Comment:
Would you consider using `switch` statement instead of multiple `if` when matching against enums? It feels more explicit.